lang/LangPrimSource/SC_Ableton_Link.cpp

```diff
@@ -0,0 +1,195 @@
+#include "PyrKernel.h"
+#include "PyrSched.h"
+#include "GC.h"
+#include "PyrPrimitive.h"
+#include "PyrSymbol.h"
+
+#include "SCBase.h"
+#include "SC_Clock.h"
+
+#ifdef SC_ABLETON_LINK
+
+#include <ableton/Link.hpp>
+
+PyrSymbol* s_LinkTempoChanged;
+
+static std::chrono::microseconds linkTimeOfInitialization;
+void initLink()
+{
+	linkTimeOfInitialization   = ableton::link::platform::Clock().micros();
+}
+
+inline std::chrono::microseconds hrToLinkTime(double secs){
+	auto time = std::chrono::duration<double>(secs);
+	return std::chrono::duration_cast<std::chrono::microseconds>(time) + linkTimeOfInitialization;
+}
+
+inline double linkToHrTime(std::chrono::microseconds micros){
+	return DurToFloat(micros - linkTimeOfInitialization);
+}
+
+class LinkClock : public TempoClock
+{
+public:
+	LinkClock(VMGlobals *inVMGlobals, PyrObject* inTempoClockObj,
+				double inTempo, double inBaseBeats, double inBaseSeconds);
+
+	~LinkClock() {}
+
+	void SetTempoAtBeat(double inTempo, double inBeats);
+	void SetTempoAtTime(double inTempo, double inSeconds);
+	void SetAll(double inTempo, double inBeats, double inSeconds);
+	double BeatsToSecs(double beats) const
+	{
+		auto timeline = mLink.captureAppTimeline();
+		double secs = linkToHrTime(timeline.timeAtBeat(beats, mQuantum));
+		return secs;
+	}
+	double SecsToBeats(double secs) const
+	{
+		auto timeline = mLink.captureAppTimeline();
+		double beats = timeline.beatAtTime(hrToLinkTime(secs), mQuantum);
+		return beats;
+	}
+
+	void SetQuantum(double quantum);
+	std::size_t NumPeers() const { return mLink.numPeers(); }
+
+protected:
+	ableton::Link mLink;
+	double mQuantum;
+};
+
+LinkClock::LinkClock(VMGlobals *inVMGlobals, PyrObject* inTempoClockObj,
+							double inTempo, double inBaseBeats, double inBaseSeconds)
+	: TempoClock(inVMGlobals, inTempoClockObj, inTempo, inBaseBeats, inBaseSeconds),
+	mLink(inTempo * 60.)
+{
+	//quantum
+	int err = slotDoubleVal(&inTempoClockObj->slots[2], &mQuantum);
+	if(err) mQuantum=4.;
```

Comment:
I'm not sure silently swallowing this error is the best approach here. Maybe the constructor should throw, and `prClock_New` can catch and report failure.